File: src/core/paths.py
"""Platform-aware ESO AddOns directory detection."""

import os
import sys
from pathlib import Path
from typing import Optional


ESO_STEAM_APP_ID = "306130"

# Relative path inside the Proton prefix to the AddOns folder
_PROTON_ADDONS_REL = Path(
    "pfx/drive_c/users/steamuser/Documents/Elder Scrolls Online/live/AddOns"
)


def _steam_library_paths() -> list[Path]:
    """Return all Steam library root paths."""
    base = Path.home() / ".local/share/Steam"
    libraries = [base]

    vdf = base / "steamapps/libraryfolders.vdf"
    if vdf.exists():
        import re
        text = vdf.read_text(errors="replace")
        for match in re.finditer(r'"path"\s+"([^"]+)"', text):
            libraries.append(Path(match.group(1)))

    return libraries


def _find_proton_addons() -> Optional[Path]:
    """Locate the ESO AddOns folder inside a Proton compatdata prefix."""
    for lib in _steam_library_paths():
        compat = lib / "steamapps/compatdata" / ESO_STEAM_APP_ID
        if compat.exists():
            return compat / _PROTON_ADDONS_REL
    return None


def detect_addons_dir() -> Optional[Path]:
    """
    Return the best-guess ESO AddOns directory for the current platform.
    Returns None if nothing is found (user will be prompted in settings).
    """
    if sys.platform == "win32":
        docs = Path.home() / "Documents"
        candidate = docs / "Elder Scrolls Online/live/AddOns"
        return candidate if candidate.exists() else docs / "Elder Scrolls Online/live/AddOns"

    # Linux / Steam Deck: try Proton prefix first, then native
    proton = _find_proton_addons()
    if proton:
        return proton

    native = Path.home() / "Documents/Elder Scrolls Online/live/AddOns"
    if native.exists():
        return native

    return None


def ensure_addons_dir(path: Path) -> None:
    path.mkdir(parents=True, exist_ok=True)

File: src-tauri/src/core/paths.rs
use std::path::{Path, PathBuf};

const ESO_STEAM_APP_ID: &str = "306130";
const PROTON_ADDONS_REL: &str = "pfx/drive_c/users/steamuser/Documents/Elder Scrolls Online/live/AddOns";

fn home() -> PathBuf {
    std::env::var("HOME").or_else(|_| std::env::var("USERPROFILE"))
        .map(PathBuf::from)
        .unwrap_or_else(|_| PathBuf::from("/"))
}

fn steam_library_paths() -> Vec<PathBuf> {
    let base = home().join(".local/share/Steam");
    let mut libs = vec![base.clone()];
    let vdf = base.join("steamapps/libraryfolders.vdf");
    if let Ok(text) = std::fs::read_to_string(&vdf) {
        let re = regex::Regex::new(r#""path"\s+"([^"]+)""#).unwrap();
        for cap in re.captures_iter(&text) {
            libs.push(PathBuf::from(&cap[1]));
        }
    }
    libs
}

fn find_proton_addons() -> Option<PathBuf> {
    for lib in steam_library_paths() {
        let compat = lib.join("steamapps/compatdata").join(ESO_STEAM_APP_ID);
        if compat.exists() {
            return Some(compat.join(PROTON_ADDONS_REL));
        }
    }
    None
}

pub fn detect_addons_dir() -> Option<String> {
    #[cfg(target_os = "windows")]
    {
        let docs = home().join("Documents");
        return Some(docs.join("Elder Scrolls Online/live/AddOns").to_string_lossy().to_string());
    }
    #[cfg(not(target_os = "windows"))]
    {
        if let Some(p) = find_proton_addons() {
            return Some(p.to_string_lossy().to_string());
        }
        let native = home().join("Documents/Elder Scrolls Online/live/AddOns");
        if native.exists() {
            return Some(native.to_string_lossy().to_string());
        }
        None
    }
}

pub fn detect_saved_vars_dir(addons_dir: Option<&str>) -> Option<String> {
    let base = if let Some(d) = addons_dir.filter(|s| !s.is_empty()) {
        Path::new(d).parent()?.to_path_buf()
    } else {
        let detected = detect_addons_dir()?;
        Path::new(&detected).parent()?.to_path_buf()
    };
    Some(base.join("SavedVariables").to_string_lossy().to_string())
}
